credentials/generate-revocation-set.py

```diff
@@ -0,0 +1,243 @@
+
+
+def extract_single_integer_attribute(subject, oid):
+    attribute_list = subject.get_attributes_for_oid(oid)
+
+    if len(attribute_list) == 1:
+        if attribute_list[0].value.isdigit():
+            return int(attribute_list[0].value)
+
+    return None
+
+
+@click.command()
+@click.help_option('-h', '--help')
+@optgroup.group('Input data sources', cls=RequiredMutuallyExclusiveOptionGroup)
+@optgroup.option('--use-main-net-dcld', type=str, default='', metavar='PATH', help="Location of `dcld` binary, to use `dcld` for mirroring MainNet.")
+@optgroup.option('--use-test-net-dcld', type=str, default='', metavar='PATH', help="Location of `dcld` binary, to use `dcld` for mirroring TestNet.")
+@optgroup.option('--use-main-net-http', is_flag=True, type=str, help="Use RESTful API with HTTPS against public MainNet observer.")
+@optgroup.option('--use-test-net-http', is_flag=True, type=str, help="Use RESTful API with HTTPS against public TestNet observer.")
+@optgroup.group('Optional arguments')
+@optgroup.option('--output', default='sample_revocation_set_list.json', type=str, metavar='FILEPATH', help="Output filename (default: sample_revocation_set_list.json)")
+def main(use_main_net_dcld, use_test_net_dcld, use_main_net_http, use_test_net_http, output):
+    """DCL PAA mirroring tools"""
+
+    production = False
+    dcld = use_test_net_dcld
+
+    if len(use_main_net_dcld) > 0:
+        dcld = use_main_net_dcld
+        production = True
+
+    use_rest = use_main_net_http or use_test_net_http
+    if use_main_net_http:
+        production = True
+
+    rest_node_url = PRODUCTION_NODE_URL_REST if production else TEST_NODE_URL_REST
+
+    if use_rest:
+        revocation_point_list = requests.get(f"{rest_node_url}/dcl/pki/revocation-points").json()["PkiRevocationDistributionPoint"]
+    else:
+        cmdlist = ['config', 'output', 'json']
+        subprocess.Popen([dcld] + cmdlist)
+
+        cmdlist = ['query', 'pki', 'all-revocation-points']
+
+        cmdpipe = subprocess.Popen(use_dcld(dcld, production, cmdlist), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        revocation_point_list = json.loads(cmdpipe.stdout.read())["PkiRevocationDistributionPoint"]
+
+    revocation_set = []
+
+    for revocation_point in revocation_point_list:
+        # 1. Validate Revocation Type
+        if revocation_point["revocationType"] != RevocationType.CRL:
+            continue
+
+        # 2. Parse the certificate
+        crl_signer_certificate = x509.load_pem_x509_certificate(revocation_point["crlSignerCertificate"])
+
+        vid = revocation_point["vid"]
+        pid = revocation_point["pid"]
+        is_paa = revocation_point["isPAA"]
+
+        # 3. && 4. Validate VID/PID
+        crl_vid = extract_single_integer_attribute(crl_signer_certificate.subject, OID_VENDOR_ID)
+        crl_pid = extract_single_integer_attribute(crl_signer_certificate.subject, OID_PRODUCT_ID)
```

Comment:
Need to support alternate representation of VID/PID (see spec "6.2.2.2. Encoding of Vendor ID and Product ID in subject and issuer fields"). Please add a TODO. The fix itself can be done in a follow-up.

Reply:
Added comments in the src file to capture the follow-up work.